File: demo-app/src/main/java/com/qualaroo/demo/SurveyDemoActivity.java
package com.qualaroo.demo;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.BottomSheetDialogFragment;
import android.support.v7.app.AppCompatActivity;
import android.text.Editable;
import android.text.TextWatcher;
import android.text.method.ScrollingMovementMethod;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.jakewharton.processphoenix.ProcessPhoenix;
import com.qualaroo.Qualaroo;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

import qualaroo.com.QualarooMobileDemo.R;

public class SurveyDemoActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        try {
            Runtime.getRuntime().exec("logcat -c");
        } catch (IOException e) {
            e.printStackTrace();
        }
        setContentView(R.layout.activity_survey);
        final Settings settings = new Settings(this);
        String apiKey = settings.getApiKey();

        final EditText apiKeyEditText = findViewById(R.id.qualaroo__demo_api_key_edit_text);
        apiKeyEditText.setText(apiKey);
        findViewById(R.id.qualaroo__demo_set_api_key_button).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                settings.storeApiKey(apiKeyEditText.getText().toString());
                ProcessPhoenix.triggerRebirth(SurveyDemoActivity.this);
            }
        });

        final View showSurveyButton = findViewById(R.id.qualaroo__demo_show_survey);
        final EditText surveyAlias = findViewById(R.id.qualaroo__demo_survey_alias_edit_text);
        surveyAlias.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void afterTextChanged(Editable editable) {
                showSurveyButton.setEnabled(editable.length() > 0);
            }
        });
        surveyAlias.setText(settings.recentSurveyAlias());
        showSurveyButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                settings.storeRecentSurveyAlias(surveyAlias.getText().toString());
                try {
                    Qualaroo.getInstance().showSurvey("mobile_test_survey");
                } catch (Exception e) {
                    Toast.makeText(getApplicationContext(), "Crash occured. Did you set a proper API key?", Toast.LENGTH_SHORT).show();
                    Log.e("Demo", e.getMessage(), e);
                }
            }
        });

        findViewById(R.id.qualaroo__demo_reset_surveys).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                getApplicationContext().deleteDatabase("qualaroo.db");
                getApplicationContext().getSharedPreferences("qualaroo_prefs", Context.MODE_PRIVATE).edit().clear().apply();
                ProcessPhoenix.triggerRebirth(SurveyDemoActivity.this);
            }
        });

        findViewById(R.id.qualaroo__demo_set_user_property).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                new AddUserPropertyDialog().show(getSupportFragmentManager(), "propertyDialog");
            }
        });

        findViewById(R.id.qualaroo__demo_check_logs).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                new LogsDialog().show(getSupportFragmentManager(), "logsDialog");
            }
        });
    }

    public static final class AddUserPropertyDialog extends BottomSheetDialogFragment {

        @Nullable
        @Override
        public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
            return inflater.inflate(R.layout.qualaroo__demo_dialog_set_user_property, container, false);
        }

        @Override
        public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
            super.onViewCreated(view, savedInstanceState);
            final EditText userPropertyName = view.findViewById(R.id.qualaroo__demo_dialog_property_name);
            final EditText userPropertyValue = view.findViewById(R.id.qualaroo__demo_dialog_property_value);

            view.findViewById(R.id.qualaroo__demo_dialog_property_confirm).setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    String name = userPropertyName.getText().toString();
                    String value = userPropertyValue.getText().toString();
                    try {
                        Qualaroo.getInstance().setUserProperty(name, value);
                    } catch (Exception e) {
                        Toast.makeText(getContext(), "Crash occured. Did you set a proper API key?", Toast.LENGTH_SHORT).show();
                        Log.e("Demo", e.getMessage(), e);
                    }
                    dismissAllowingStateLoss();
                }
            });
            view.findViewById(R.id.qualaroo__demo_dialog_property_cancel).setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View view) {
                            dismissAllowingStateLoss();
                        }
                    }
            );
        }
    }

    public static final class LogsDialog extends BottomSheetDialogFragment {

        private TextView logsView;

        @Nullable
        @Override
        public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
            return inflater.inflate(R.layout.qualaroo__demo_dialog_logs, container, false);
        }

        @Override
        public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
            super.onViewCreated(view, savedInstanceState);
            logsView = view.findViewById(R.id.qualaroo__demo_dialog_logs);
            logsView.setMovementMethod(new ScrollingMovementMethod());
            view.findViewById(R.id.qualaroo__demo_dialog_logs_close).setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    dismissAllowingStateLoss();
                }
            });

            view.findViewById(R.id.qualaroo__demo_dialog_logs_copy).setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    ClipboardManager clipboard = (ClipboardManager) getContext().getSystemService(Context.CLIPBOARD_SERVICE);
                    if (clipboard != null) {
                        ClipData clip = ClipData.newPlainText("logcat", logsView.getText().toString());
                        clipboard.setPrimaryClip(clip);
                        Toast.makeText(getContext(), "Logs copied to clipboard", Toast.LENGTH_SHORT).show();
                    } else {
                        Toast.makeText(getContext(), "Could not copy logs", Toast.LENGTH_SHORT).show();
                    }
                }
            });

            populateLogs();
        }

        private void populateLogs() {
            try {
                Process process = Runtime.getRuntime().exec("logcat -d -v long QualarooSDK:V");
                BufferedReader bufferedReader = new BufferedReader(
                        new InputStreamReader(process.getInputStream()));

                StringBuilder log=new StringBuilder();
                String line;

                while ((line = bufferedReader.readLine()) != null) {
                    log.append(line);
                    log.append('\n');
                }
                logsView.setText(log.toString());
            } catch (IOException e) {
                //ignore
            }

        }
    }
}

File: demo-app/src/main/java/com/qualaroo/demo/SurveyDemoActivity.kt
/*
 * Copyright (c) 2018, Qualaroo, Inc. All Rights Reserved.
 *
 * Please refer to the LICENSE.md file for the terms and conditions
 * under which redistribution and use of this file is permitted.
 */

package com.qualaroo.demo

import android.content.Context
import android.os.Bundle
import androidx.localbroadcastmanager.content.LocalBroadcastManager

//import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import com.jakewharton.processphoenix.ProcessPhoenix
import com.qualaroo.Qualaroo
import com.qualaroo.QualarooSurveyEventReceiver
import com.qualaroo.demo.dialog.AddUserPropertyDialog
import com.qualaroo.demo.dialog.LogsDialog
import com.qualaroo.demo.repository.SurveyAliasesRepository
import com.qualaroo.demo.util.Logcat
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import qualaroo.com.QualarooMobileDemo.R

class SurveyDemoActivity : AppCompatActivity() {

    private val compositeDisposable = CompositeDisposable()
    private val aliasSpinner by lazy { findViewById<Spinner>(R.id.qualaroo__demo_survey_alias_spinner) }
    private val aliasSpinnerProgressBar by lazy { findViewById<ProgressBar>(R.id.qualaroo__demo_survey_alias_spinner_progress_bar) }
    private val showButton by lazy { findViewById<View>(R.id.qualaroo__demo_show_survey) }

    private var aliasAdapter: ArrayAdapter<String>? = null
    private lateinit var surveyAliasesRepository: SurveyAliasesRepository
    private val broadcastReceiver = object : QualarooSurveyEventReceiver() {
        override fun onSurveyEvent(surveyAlias: String, eventType: Int) {
            when (eventType) {
                EVENT_TYPE_SHOWN -> Log.d("Observer", "$surveyAlias has been shown")
                EVENT_TYPE_DISMISSED -> Log.d("Observer", "$surveyAlias has been dismissed")
                EVENT_TYPE_FINISHED -> Log.d("Observer", "$surveyAlias has been finished")
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Logcat.clearLogcat()
        setContentView(R.layout.activity_survey)
        val settings = Settings(this)

        val apiKey = settings.apiKey()
        val apiKeyEditText = findViewById<EditText>(R.id.qualaroo__demo_api_key_edit_text)
        apiKeyEditText.setText(apiKey)

        surveyAliasesRepository = DependenciesComponent(apiKey).surveysRepository()

        findViewById<View>(R.id.qualaroo__demo_set_api_key_button).setOnClickListener {
            settings.storeApiKey(apiKeyEditText.text.toString())
            ProcessPhoenix.triggerRebirth(this@SurveyDemoActivity)
        }

        aliasAdapter = ArrayAdapter(this, android.R.layout.simple_list_item_1)
        aliasSpinner.adapter = aliasAdapter
        showButton.setOnClickListener {
            val alias = aliasSpinner.selectedItem as String
            settings.storeRecentSurveyAlias(alias)
            showSurvey(alias)
        }
        findViewById<View>(R.id.qualaroo__demo_reset_surveys).setOnClickListener { clearAndRestartApp() }
        findViewById<View>(R.id.qualaroo__demo_set_user_property).setOnClickListener { showAddUserPropertyDialog() }
        findViewById<View>(R.id.qualaroo__demo_check_logs).setOnClickListener { showLogsDialog() }

        loadAvailableAliases()
        LocalBroadcastManager.getInstance(this).registerReceiver(broadcastReceiver, QualarooSurveyEventReceiver.intentFilter())
    }

    private fun loadAvailableAliases() {
        aliasSpinner.hide()
        showButton.hide()
        aliasSpinnerProgressBar.show()
        val disposable = surveyAliasesRepository.fetchAliases()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ populateAdapter(it) }, { handleError(it) })
        compositeDisposable.add(disposable)
    }

    private fun populateAdapter(aliases: List<String>) {
        aliasSpinnerProgressBar.hide()
        aliasSpinner.show()
        showButton.show()
        aliasAdapter?.addAll(aliases)
    }

    private fun handleError(throwable: Throwable) {
        Log.e("Demo", "Exception caught!", throwable)
    }

    private fun showSurvey(alias: String) {
        try {
            Qualaroo.getInstance().showSurvey(alias)
        } catch (e: Exception) {
            Toast.makeText(applicationContext, "Crash occured. Did you set a proper API key?", Toast.LENGTH_SHORT).show()
            Log.e("Demo", e.message, e)
        }
    }

    private fun showLogsDialog() {
        LogsDialog().show(supportFragmentManager, "logsDialog")
    }

    private fun showAddUserPropertyDialog() {
        AddUserPropertyDialog().show(supportFragmentManager, "propertyDialog")
    }

    private fun clearAndRestartApp() {
        applicationContext.deleteDatabase("qualaroo.db")
        applicationContext.getSharedPreferences("qualaroo_prefs", Context.MODE_PRIVATE).edit().clear().apply()
        ProcessPhoenix.triggerRebirth(this@SurveyDemoActivity)
    }

    private fun View.show() {
        visibility = View.VISIBLE
    }

    private fun View.hide() {
        visibility = View.GONE
    }

    override fun onDestroy() {
        LocalBroadcastManager.getInstance(this).unregisterReceiver(broadcastReceiver)
        super.onDestroy()
    }
}
